rclpy/src/rclpy/client.cpp

```diff
@@ -0,0 +1,219 @@
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "_rclpy_destroy_client got NULL pointer");
+    return;
+  }
+
+  rcl_ret_t ret = rcl_client_fini(&(cli->client), cli->node);
+  if (RCL_RET_OK != ret) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to fini client: %s",
+      rcl_get_error_string().str);
+    rcl_reset_error();
+  }
+  PyMem_Free(cli);
+}
+
+py::capsule
+client_create(
+  py::capsule pynode, py::object pysrv_type, std::string service_name,
+  py::capsule pyqos_profile)
+{
+  auto node = static_cast<rcl_node_t *>(
+    rclpy_handle_get_pointer_from_capsule(pynode.ptr(), "rcl_node_t"));
+  if (!node) {
+    throw py::error_already_set();
+  }
+
+  auto srv_type = static_cast<rosidl_service_type_support_t *>(
+    rclpy_common_get_type_support(pysrv_type.ptr()));
+  if (!srv_type) {
+    throw py::error_already_set();
+  }
+
+  rcl_client_options_t client_ops = rcl_client_get_default_options();
+
+  if (!pyqos_profile.is_none()) {
+    if (0 != strcmp("rmw_qos_profile_t", pyqos_profile.name())) {
+      throw py::value_error("capsule is not an rmw_qos_profile_t");
+    }
+    auto qos_profile = static_cast<rmw_qos_profile_t *>(pyqos_profile);
+    client_ops.qos = *qos_profile;
+  }
+
+  // Use smart pointer to make sure memory is free'd on error
+  auto deleter = [](rclpy_client_t * ptr) {_rclpy_destroy_client(ptr);};
+  auto cli = std::unique_ptr<rclpy_client_t, decltype(deleter)>(
+    static_cast<rclpy_client_t *>(PyMem_Malloc(sizeof(rclpy_client_t))),
+    deleter);
+  if (!cli) {
+    throw std::bad_alloc();
+  }
+  cli->client = rcl_get_zero_initialized_client();
+  cli->node = node;
+
+  rcl_ret_t ret = rcl_client_init(
+    &(cli->client), node, srv_type,
+    service_name.c_str(), &client_ops);
+  if (ret != RCL_RET_OK) {
+    if (ret == RCL_RET_SERVICE_NAME_INVALID) {
+      std::string error_text{"failed to create client due to invalid service name '"};
+      error_text += service_name;
+      error_text += "': ";
+      error_text += rcl_get_error_string().str;
+      rcl_reset_error();
+      throw py::value_error(error_text);
+    }
+    throw RCLError("failed to create client");
+  }
+
+  PyObject * pycli_c =
+    rclpy_create_handle_capsule(cli.get(), "rclpy_client_t", _rclpy_destroy_client);
+  if (!pycli_c) {
+    throw py::error_already_set();
+  }
+  auto pycli = py::reinterpret_steal<py::capsule>(pycli_c);
+  // pycli now owns the rclpy_client_t
+  cli.release();
+
+  auto cli_handle = static_cast<rclpy_handle_t *>(pycli);
+  auto node_handle = static_cast<rclpy_handle_t *>(pynode);
+  _rclpy_handle_add_dependency(cli_handle, node_handle);
+  if (PyErr_Occurred()) {
+    _rclpy_handle_dec_ref(cli_handle);
+    throw py::error_already_set();
+  }
+
+  return pycli;
+}
+
+uint64_t
```

Comment:
The [return type should be `int64_t` to match `rcl_send_request()`](https://github.com/ros2/rcl/blob/4740c82864518a331ae98799f25b2ba085b22473/rcl/include/rcl/client.h#L248)

Reply:
Ah, good catch. Done in 97b35b8.